pageserver/src/layered_repository.rs

```diff
@@ -0,0 +1,833 @@
+        self.checkpoint()?;
+
+        LayeredRepository::gc_iteration(self.conf, horizon)
+    }
+
+    fn put_wal_record(&self, tag: BufferTag, rec: WALRecord) -> Result<()> {
+        debug!("put_wal_record: {:?} at {}", tag, rec.lsn);
+
+        let snapfile = self.get_snapshot_file_for_write(tag.rel, rec.lsn)?;
+        snapfile.put_wal_record(tag.blknum, rec)
+    }
+
+    fn put_truncation(&self, rel: RelTag, lsn: Lsn, relsize: u32) -> anyhow::Result<()> {
+        debug!("put_truncation: {:?} at {}", relsize, lsn);
+
+        let snapfile = self.get_snapshot_file_for_write(rel, lsn)?;
+        snapfile.put_truncation(lsn, relsize)
+    }
+
+    fn put_page_image(&self, tag: BufferTag, lsn: Lsn, img: Bytes) -> Result<()> {
+        debug!("put_page_image: {:?} at {}", tag, lsn);
+
+        let snapfile = self.get_snapshot_file_for_write(tag.rel, lsn)?;
+        snapfile.put_page_image(tag.blknum, lsn, img)
+    }
+
+    fn put_unlink(&self, rel: RelTag, lsn: Lsn) -> Result<()> {
+        debug!("put_unlink: {} at {}", rel, lsn);
+
+        let snapfile = self.get_snapshot_file_for_write(rel, lsn)?;
+        snapfile.put_unlink(lsn)
+    }
+
+    ///
+    /// Flush to disk all data that was written with the put_* functions
+    ///
+    /// NOTE: This has nothing to do with checkpoint in PostgreSQL. We don't
+    /// know anything about them here in the repository.
+    fn checkpoint(&self) -> Result<()> {
+        let last_valid_lsn = self.last_valid_lsn.load();
+
+        let mut layers = self.layers.lock().unwrap();
+
+        // Walk through each in-memory, and write any dirty data to disk,
+        // as snapshot files.
+        //
+        // We currently write a new snapshot file for every relation
+        // that was modified, if there has been any changes at all.
+        // It would be smarter to only flush out in-memory layers that
+        // have accumulated a fair amount of changes. Note that the
+        // start and end LSNs of snapshot files belonging to different
+        // relations don't have to line up, although currently they do
+        // because of the way this works. So you could have a snapshot
+        // file covering LSN range 100-200 for one relation, and a
+        // snapshot file covering 150-250 for another relation. The
+        // read functions should even cope with snapshot files
+        // covering overlapping ranges for the same relation, although
+        // that situation never arises currently.
+        //
+        // Note: We release all the layer structs, and start fresh
+        // with an empty map. This keeps memory usage in check, but is
+        // probably too aggressive. Some kind of LRU policy would be
+        // appropriate.
+        //
+        let snapfiles = std::mem::take(&mut *layers);
+        for snapfile in snapfiles.0.values() {
+            if !snapfile.is_frozen() {
+                snapfile.freeze(last_valid_lsn + 1)?;
```

Comment:
I wonder why it is needed to add 1 to last_valid_lsn here?

Reply:
I've been sloppy on whether the start and end bounds are inclusive or exclusive. This is one such instance. TODO: make sure we're consistent on that, check if this + 1 still makes sense and remove or add a comment.

Reply:
Fixed this.